src/backend/distributed/transaction/lock_graph.c

```diff
@@ -0,0 +1,752 @@
+	int blockingNodeId;
+	int64 blockingTransactionNum;
+	TimestampTz blockingTransactionStamp;
+
+	/* blocking transaction is also waiting on a lock */
+	bool isBlockingXactWaiting;
+} WaitEdge;
+
+
+/*
+ * WaitGraph represent a graph of wait edges as an adjacency list.
+ */
+typedef struct WaitGraph
+{
+	int localNodeId;
+	size_t numAllocated;
+	size_t numUsed;
+	WaitEdge *edges;
+} WaitGraph;
+
+
+/*
+ * PROCStack is a stack of PGPROC pointers used to perform a depth-first search
+ * through the lock graph.
+ */
+typedef struct PROCStack
+{
+	int procCount;
+	PGPROC **procs;
+} PROCStack;
+
+
+static void AddRemoteWaitEdges(WaitGraph *waitGraph);
+static void AddWaitEdgeFromResult(WaitGraph *waitGraph, PGresult *result, int rowIndex);
+static int64 ParseIntField(PGresult *result, int rowIndex, int colIndex);
+static bool ParseBoolField(PGresult *result, int rowIndex, int colIndex);
+static TimestampTz ParseTimestampTzField(PGresult *result, int rowIndex, int colIndex);
+static void ReturnWaitGraph(WaitGraph *waitGraph, FunctionCallInfo fcinfo);
+static WaitGraph * BuildWaitGraphForSourceNode(int sourceNodeId);
+static void LockLockData(void);
+static void UnlockLockData(void);
+static void AddEdgesForLockWaits(WaitGraph *waitGraph, PGPROC *waitingProc,
+								 PROCStack *remaining);
+static void AddEdgesForWaitQueue(WaitGraph *waitGraph, PGPROC *waitingProc,
+								 PROCStack *remaining);
+static void AddWaitEdge(WaitGraph *waitGraph, BackendData *waitingXactData,
+						PGPROC *waitingProc, BackendData *blockingXactData,
+						PGPROC *blockingProc, bool isBlockingProcessWaiting);
+static WaitEdge * AllocWaitEdge(WaitGraph *waitGraph);
+static bool IsProcessWaitingForLock(PGPROC *proc);
+static bool IsSameLockGroup(PGPROC *leftProc, PGPROC *rightProc);
+
+
+PG_FUNCTION_INFO_V1(dump_local_wait_edges);
+PG_FUNCTION_INFO_V1(dump_global_wait_edges);
+
+
+/*
+ * dump_global_wait_edges returns global wait edges for distributed transactions
+ * originating from the node on which it is started.
+ */
+Datum
+dump_global_wait_edges(PG_FUNCTION_ARGS)
+{
+	int32 localNodeId = 0;
+	WaitGraph *waitGraph = NULL;
+
+	localNodeId = GetLocalGroupId();
+	waitGraph = BuildWaitGraphForSourceNode(localNodeId);
+
+	AddRemoteWaitEdges(waitGraph);
+
+	ReturnWaitGraph(waitGraph, fcinfo);
+
+	return (Datum) 0;
+}
+
+
+/*
+ * AddRemoteWaitEdges adds all wait edges on (other) worker nodes to the given
+ * graph.
+ */
+static void
+AddRemoteWaitEdges(WaitGraph *waitGraph)
+{
+	List *workerNodeList = ActiveWorkerNodeList();
+	ListCell *workerNodeCell = NULL;
+	char *nodeUser = CitusExtensionOwnerName();
+	List *connectionList = NIL;
+	ListCell *connectionCell = NULL;
+
+	/* open connections in parallel */
+	foreach(workerNodeCell, workerNodeList)
+	{
+		WorkerNode *workerNode = (WorkerNode *) lfirst(workerNodeCell);
+		char *nodeName = workerNode->workerName;
+		int nodePort = workerNode->workerPort;
+		MultiConnection *connection = NULL;
+		int connectionFlags = 0;
```

Comment:
Then, shall we use `FORCE_NEW_CONNECTION` to make this behavior consistent with every usage pattern? 

Because currently we may or may not open a new connection depending on the cached connections.